packages/edit-site/src/store/actions.js

```diff
@@ -228,3 +230,83 @@ export function updateSettings( settings ) {
 		settings,
 	};
 }
+
+/**
+ * Reverts a template to its original theme-provided file.
+ *
+ * This works with the assumption that the template being reverted
+ * belongs to the current theme.
+ *
+ * @param {Object} template The template to revert.
+ */
+export function* revertTemplate( template ) {
+	const postType = yield controls.resolveSelect(
+		'core',
+		'getPostType',
+		'wp_template'
+	);
+
+	const currentTheme = yield controls.resolveSelect(
+		'core',
+		'getCurrentTheme'
+	);
+
+	if ( ! isTemplateRevertable( template, currentTheme?.stylesheet ) ) {
+		yield controls.dispatch(
+			'core/notices',
+			'createErrorNotice',
+			__( 'This template is not revertable' ),
+			{ type: 'snackbar' }
+		);
+		return;
+	}
+
+	try {
+		yield apiFetch( {
+			path: `/wp/v2/${ postType.rest_base }/${ template.id }`,
+			method: 'DELETE',
+		} );
+
+		const fileTemplates = yield controls.resolveSelect(
+			'core',
+			'getEntityRecords',
+			'postType',
+			'wp_template',
+			{
+				slug: template.slug,
+				status: 'auto-draft',
+				per_page: 1,
+				theme: currentTheme?.stylesheet,
+			}
+		);
+
+		if ( ! fileTemplates.length ) {
+			yield controls.dispatch(
+				'core/notices',
+				'createErrorNotice',
+				__( 'This template is not revertable' ),
+				{ type: 'snackbar' }
+			);
+			return;
+		}
```

Comment:
It looks like we can create this notice in 2 places:
1. Prior to this step when `isTemplateRevertable` comes back falsy.
2. Here after we have already sent the `apiFetch` to delete the template and fail to find the auto-draft to replace it with.

Should we be checking that autodraft and conditionally sending this notice before sending the `apiFetch` to delete?

Reply:
Oh, Yeah I see why we have to do it this way.. 🤔   and missed your recent note above:

>Now, the case 2 is not really testable, and would cause many headaches if it happens for real. Given how the sync mechanism work, it shouldn't be possible to happen, but I've still decided to add an error notice there just in case.
Maybe the notice could change into "An unexpected error occurred. Please reload."

The unexpected error message may make more sense in this case. 🤔   

Perhaps it would make sense to always create the auto-draft if the auto-draft doesn't exist.  So the new auto-draft would be created as soon as the original one was customized and published, as opposed to right after it is deleted?  Not a blocker for this PR, but a follow up if necessary.